mhs/mhs/common/state/work_description.py

```diff
@@ -0,0 +1,122 @@
+import json
+import utilities.integration_adaptors_logger as log
+
+
+logger = log.IntegrationAdaptorsLogger('STATE_MANAGER')
+
+
+class MessageStatus:
+    RECEIVED = 1
+    STARTED = 2
+    IN_OUTBOUND_WORKFLOW = 3
+
+
+DATA_KEY = 'MESSAGE_KEY'
+VERSION_KEY = 'VERSION'
+TIMESTAMP = 'TIMESTAMP'
+DATA = 'DATA'
+STATUS = 'STATUS'
+
+
+class OutOfDateVersionError(RuntimeError):
+    pass
+
+
+class EmptyWorkDescriptionError(RuntimeError):
+    pass
+
+
+class WorkDescription:
+
+    def __init__(self, persistence_store, table_name: str, store_data: dict):
+        if persistence_store is None:
+            raise ValueError('Expected persistence store')
+        if not table_name:
+            raise ValueError('Expected non empty table name')
+
+        self.persistence_store = persistence_store
+        self.table_name = table_name
+
+        data = store_data[DATA]
+        self.message_key = store_data[DATA_KEY]
+        self.version = data[VERSION_KEY]
+        self.timestamp = data[TIMESTAMP]
+        self.status = data[STATUS]
+
+    async def publish(self):
+        logger.info('011', 'Attempting to publish work description {key}', {'key': self.message_key})
+        logger.info('012', 'Retrieving latest work description to check version')
+        latest_data = await self.persistence_store.get(self.table_name, self.message_key)
+        if latest_data is not None:
+            latest_version = latest_data[DATA][VERSION_KEY]
+            if latest_version > self.version:
+                logger.error('013', 'Failed to update message {key}, local version out of date',
+                             {'key': self.message_key})
+                raise OutOfDateVersionError(f'Failed to update message {self.message_key}: local version out of date ')
+
+        serialised = self._serialise_data()
+
+        old_data = await self.persistence_store.add(self.table_name, serialised)
+        logger.info('014', 'Successfully updated work description to state store for {key}', {'key': self.message_key})
+        return old_data
+
+    def _serialise_data(self):
+        data = {
+            DATA_KEY: self.message_key,
+            'DATA': {
+                TIMESTAMP: self.timestamp,
+                VERSION_KEY: self.version,
+                STATUS: self.status
+            }
+        }
+
+        return json.dumps(data)
+
+
+class WorkDescriptionFactory:
```

Comment:
Yeah that's a good point, I can move them out of the class but does this raise the question as to if we should change that import rule?

Reply:
@flavgj @dshipp-bjss any thoughts?

Reply:
I’d envisaged that they’d be static members of the WorkDescription class rather than in their own class. They could also be stand alone functions - I’m not precious

Reply:
I'll make them as standalone functions simply since the code guidelines say to try and avoid static methods where possible